pyrolite/mineral/normative.py

```diff
@@ -301,308 +302,661 @@ def CIPW_norm(df, Fe_correction=None, adjust_all=False):
         )
 
     df = df.reindex(columns=columns).fillna(0)
-    ############################################################################
-    # Verma's ADJ columns are equivalent to `df_update.pyrocomp.renormalise()`
-    # at this point in the workflow.
-    ############################################################################
-    # Normalise to 100 molar percent on anhydrous basis
-    # res = df[columns].pyrochem.to_molecular()
-    res = df.div(pd.Series([pt.formula(c).mass for c in columns], index=columns))
-    res = res.divide(res.sum(axis=1) / 100, axis=0).fillna(0)
-
-    # endmember component fractions
-    res.pyrochem.add_MgNo()
-    MgNo, FeNo = res["Mg#"], 1 - res["Mg#"]
-    xFeO = res["FeO"] / (res["FeO"] + res["MnO"])
-    xMnO = 1 - xFeO
-    ############################################################################
-    # Calculate effective molecular weights for silicate & oxide endmembers
-    ############################################################################
-    # When updating for trace elements will need to add more here
-    mineral_mw = {}
-
-    # Components (used only for calc) ##########################################
-    # effective MW of (Fe2+, Mn2+)O
-    molw_FeO = xMnO * pt.formula("MnO").mass + xFeO * pt.formula("FeO").mass
-    # effective MW of Mn-fayalite
-    molw_fayalite = (2 * molw_FeO) + pt.formula("SiO2").mass
-    # effective v of ferrosilite (Fe2+, Mn2+)SiO3
-    molw_fs = molw_FeO + pt.formula("SiO2").mass
-    # Endmembers (used in the norm) ############################################
-    # effective MW of ilmenite (Fe2+, Mn2+)TiO3
-    molw_il = molw_FeO + pt.formula("TiO2").mass
-    # effective MW of hypersthene (Mg2+, Fe2+, Mn2+)SiO3
-    molw_hyp = pt.formula("MgSiO3").mass * MgNo + molw_fs * FeNo
-    # effective MW of olivine (Mg2+, Fe2+, Mn2+)SiO4
-    molw_ol = pt.formula("Mg2SiO4").mass * MgNo + molw_fayalite * FeNo
-    # effective MW of magnetite
-    molw_mt = molw_FeO + pt.formula("Fe2O3").mass
-
-    for n, mw in zip(
-        ["ilmenite", "hypersthene", "olivine", "magnetite"],
-        [molw_il, molw_hyp, molw_ol, molw_mt],
-    ):
-        mineral_mw[n] = mw
-    # anhydrous apatite (Ca10P6O25)/3 relative to P2O5
-    mineral_mw["apatite"] = (
-        pt.formula("Ca10(PO4)6(OH)2").mass - pt.formula("H2O").mass
-    ) / 3
-    mineral_mw["halite"] = pt.formula("NaCl").mass  # leave out Cl?
-    mineral_mw["fluorite"] = pt.formula("CaF2").mass  # leave out F2?
-    mineral_mw["pyrite"] = pt.formula("FeS2").mass
-    ############################################################################
-    # Aggregate Mn, Fe
-    ############################################################################
-    res["FeO"] = res["FeO"] + res["MnO"]
-    res = res.drop(["MnO"], axis=1)
+
+
+    majors = ['SiO2', 'TiO2', 'Al2O3', 'Fe2O3', 'FeO', 'MnO', 'MgO', 'CaO',
+       'Na2O', 'K2O', 'P2O5']
+
+    element_AW = {
```

Comment:
@morganjwilliams I think some of these dicts of values can be cleaned up using periodictable if need be